File: synara/src/app/features/matrix-dto/thread.ts
/**
 * Thread summary DTO.
 */

import type { EventId, RoomId } from './ids';
import {
  hasForbiddenWireFields,
  isObject,
  optNumber,
  optString,
  reqBoolean,
  reqNumber,
  reqString,
} from './parseUtil';

export type ThreadSummary = {
  roomId: RoomId;
  rootEventId: EventId;
  replyCount: number;
  latestEventId?: EventId;
  latestOriginServerTs?: number;
  participated: boolean;
  unreadCount?: number;
};

export function parseThreadSummary(value: unknown): ThreadSummary | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const roomId = reqString(value, 'roomId');
  const rootEventId = reqString(value, 'rootEventId');
  const replyCount = reqNumber(value, 'replyCount');
  const latestEventId = optString(value, 'latestEventId');
  const latestOriginServerTs = optNumber(value, 'latestOriginServerTs');
  const participated = reqBoolean(value, 'participated');
  const unreadCount = optNumber(value, 'unreadCount');
  if (
    roomId === null ||
    rootEventId === null ||
    replyCount === null ||
    latestEventId === null ||
    latestOriginServerTs === null ||
    participated === null ||
    unreadCount === null
  ) {
    return null;
  }
  return {
    roomId,
    rootEventId,
    replyCount,
    latestEventId,
    latestOriginServerTs,
    participated,
    unreadCount,
  };
}

File: crates/synara-core/src/dto/thread.rs
//! Thread summary DTO.

use serde::{Deserialize, Serialize};

use super::ids::{EventId, RoomId};

/// Thread root summary for timeline / list UI.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThreadSummary {
    pub room_id: RoomId,
    pub root_event_id: EventId,
    pub reply_count: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latest_event_id: Option<EventId>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latest_origin_server_ts: Option<u64>,
    pub participated: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unread_count: Option<u32>,
}
